05_debug/procfs_uppercase_conv/khmod.c

```diff
@@ -0,0 +1,134 @@
+// SPDX-License-Identifier: MIT and GPL
+
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/device.h>
+#include <linux/err.h>
+#include <linux/list.h>
+#include <linux/slab.h>
+#include <linux/ctype.h>
+#include <linux/proc_fs.h>
+#include <linux/uaccess.h>
+
+#define DIR_NAME	"kharchuk"
+#define ENT1_NAME	"rw"
+#define ENT2_NAME	"stats"
+
+MODULE_DESCRIPTION("Sysfs lowercase converter");
+MODULE_AUTHOR("vadym.kharchuk@globallogic.com");
+MODULE_VERSION("0.1");
+MODULE_LICENSE("Dual MIT/GPL");
+
+static struct proc_dir_entry *dir;
+static struct proc_dir_entry *ent1, *ent2;
+
+static char message[PAGE_SIZE];
+
+struct stats {
+	int total_calls;
+	int char_processed;
+	int char_converted;
+};
+
+struct stats toupper_stats;
+ssize_t msg_size;
+
+static void string_toupper(char *msg)
+{
+	int i;
+
+	for (i = 0; msg[i]; i++) {
+		if (islower(msg[i])) {
+			msg[i] = toupper(msg[i]);
+			toupper_stats.char_converted++;
+		}
+	}
+
+	toupper_stats.total_calls++;
+	toupper_stats.char_processed += strlen(msg);
+}
+
+static ssize_t main_write(struct file *file, const char __user *pbuf,
+			size_t count, loff_t *ppos)
+{
+	ssize_t not_copied;
+
+	not_copied = copy_from_user(message, pbuf, count);
+	msg_size = count - not_copied;
+	string_toupper(message);
+
+	return msg_size;
+}
+
+static ssize_t main_read(struct file *file, char __user *pbuf,
+			size_t count, loff_t *ppos)
+{
+	return simple_read_from_buffer(pbuf, count, ppos,
+					message, strlen(message));
+}
+
+static ssize_t stats_read(struct file *file, char __user *pbuf,
+			size_t count, loff_t *ppos)
+{
+	char report[255];
+
+	sprintf(report, "Total calls - %d, characters processed - %d characters converted - %d\n",
```

Comment:
Please create "define" for this buffer length constant.
Better to use snprintf instead of sprintf.